src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerExecutor.cs

```diff
@@ -1238,7 +1233,7 @@ private Action<Diagnostic> GetAddDiagnostic(ISymbol contextSymbol, ImmutableArra
         {
             return diagnostic =>
             {
-                if (isGeneratedCodeSymbol && shouldSuppressGeneratedCodeDiagnostic(diagnostic, analyzer, compilation, cancellationToken))
+                if (shouldSuppressGeneratedCodeDiagnostic(diagnostic, analyzer, compilation, cancellationToken))
```

Comment:
the cache it does right now is only partial I think. once it ever determined a tree has no generated code, it should be cheap after that.

also, if there is a case where the flag couldn't determine whether it is generated or not, how about make the flag tri state and do only expensive call path when the flag is un-derministic. I believe only place where it is un-determistic is when analyzer is compilation end analyzer and that is rare case. so rather than exchange common case cheap and rare case expensive to common case expensive and rare case relatively cheap, we only make the rare case a bit more expensive (with some more optimization so that even that can be a bit more cheap)